### ISSUE TITLE
caret does not use the master node of the PSOCKcluster when using parallel backend

### ISSUE THREAD

**Issue by fgnu** [topepo/caret#322]
**Cross-posted from [here](http://stackoverflow.com/questions/33718620/r-caret-does-not-use-the-host-of-psockcluster-when-using-parallel-backend).**

I am trying to get `caret` to train `xgboost` models over a grid of hyperparameters using a parallel backend. 

Here is some code that uses the [*Give Me Some Credit*][1] data to demonstrate setting up a parallel backend for `caret`'s hyperparameter grid search. 

    library(plyr)
    library(dplyr)
    library(pROC)
    library(caret)
    library(xgboost)
    library(readr)
    library(parallel)
    library(doParallel)
    
    if(exists("xgboost_cluster")) stopCluster(xgboost_cluster)
    hosts = paste0("192.168.18.", 52:53)
    xgboost_cluster = makePSOCKcluster(hosts, master="192.168.18.51")
    
    # load the packages across the cluster
    clusterEvalQ(xgboost_cluster, {
      deps = c("plyr", "Rcpp", "dplyr", "caret", "xgboost", "pROC", "foreach", "doParallel")
      for(d in deps) library(d, character.only = TRUE)
      rm(d, deps)
    })
    
    registerDoParallel(xgboost_cluster)  
    # load in the training data
    df_train = read_csv("04-GiveMeSomeCredit/Data/cs-training.csv") %>%
      na.omit() %>%                                                                # listwise deletion 
      select(-`[EMPTY]`) %>%
      mutate(SeriousDlqin2yrs = factor(SeriousDlqin2yrs,                           # factor variable for classification
                                       labels = c("Failure", "Success")))
    # set up the cross-validated hyper-parameter search
    xgb_grid_1 = expand.grid(
      nrounds = 1000,
      eta = c(0.01, 0.001, 0.0001),
      max_depth = c(2, 4, 6, 8, 10),
      gamma = 1
    )
    
    # pack the training control parameters
    xgb_trcontrol_1 = trainControl(
      method = "cv",
      number = 5,
      verboseIter = TRUE,
      returnData = FALSE,
      returnResamp = "all",                                                        # save losses across all models
      classProbs = TRUE,                                                           # set to TRUE for AUC to be computed
      summaryFunction = twoClassSummary,
      allowParallel = TRUE
    )
    
    # train the model for each parameter combination in the grid, 
    #   using CV to evaluate
    xgb_train_1 = train(
      x = as.matrix(df_train %>%
                      select(-SeriousDlqin2yrs)),
      y = as.factor(df_train$SeriousDlqin2yrs),
      trControl = xgb_trcontrol_1,
      tuneGrid = xgb_grid_1,
      method = "xgbTree"
    )

I checked that all cores on the `hosts` are being utilized for training, but on the `master` node, no processes are utilized. Is this expected behaviour? Any way to change this behaviour and leverage the cores on the master node for processing as well?

  [1]: https://www.kaggle.com/c/GiveMeSomeCredit/data

**Comment by topepo** [comment]
It would be good to have the results of `sesisonInfo()`.

Can you run this and give the results?

```r
library(plyr)
library(dplyr)
library(pROC)
library(caret)
library(xgboost)
library(readr)
library(parallel)
library(doParallel)

library(foreach)

if(exists("xgboost_cluster")) stopCluster(xgboost_cluster)
hosts = paste0("192.168.18.", 52:53)
xgboost_cluster = makePSOCKcluster(hosts, master="192.168.18.51")

# load the packages across the cluster
clusterEvalQ(xgboost_cluster, {
  deps = c("plyr", "Rcpp", "dplyr", "caret", "xgboost", "pROC", "foreach", "doParallel")
  for(d in deps) library(d, character.only = TRUE)
  rm(d, deps)
})

registerDoParallel(xgboost_cluster)  

foreach(i = 1:3) %dopar% {
  library(caret)
  train(Species ~ ., data = iris, method = "knn")
}
```

`caret` uses a pretty straightforward application of `foreach`. If you are having issues, it is most likely due to how you've set things up. 

Also, please show that the code works sequentially too.

**Comment by fgnu** [comment]
@topepo Obviously, as you said, this had nothing to do with `caret` -- thanks for your reply nonetheless. [This answer](http://stackoverflow.com/a/33740500/1414455) on SO helped me solve this.
